### PR TITLE
Feature/devise user

### PR DESCRIPTION
**Pull request by apeksha2697**
This feature was needed to create authentication for user.

**Changes**
1. Added devise to user
2. Added navbar and side panel
3. created signup and log in form
4. created view for all members in a company
5. created users controller
6. created view for dashboard

### CODE REVIEW

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class HomeController < ApplicationController
+  before_action :authenticate_user!
+  def dashboard
```
Comment (BandanaPandey) [inline app/controllers/home_controller.rb:4]: add one line space between before_action and dashboard action @apeksha2697

File: app/controllers/users/unlocks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
```
Comment (BandanaPandey) [inline app/controllers/users/unlocks_controller.rb]: @apeksha2697  please remove this file from PR

File: app/controllers/users/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
```
Comment (BandanaPandey) [inline app/controllers/users/sessions_controller.rb]: @apeksha2697  please remove this file from PR

File: app/controllers/users/registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+class Users::RegistrationsController < Devise::RegistrationsController
```
Comment (BandanaPandey) [inline app/controllers/users/registrations_controller.rb:1]: @apeksha2697  please remove commented code from this file.
Ideally your PR should never contain commented code

File: app/controllers/users/passwords_controller.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class Users::PasswordsController < Devise::PasswordsController
+  # GET /resource/password/new
```
Comment (BandanaPandey) [inline app/controllers/users/passwords_controller.rb]: @apeksha2697  please remove this file from PR

File: app/controllers/users/omniauth_callbacks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  # You should configure your model like this:
```
Comment (BandanaPandey) [inline app/controllers/users/omniauth_callbacks_controller.rb]: @apeksha2697  please remove this file from PR

File: app/controllers/users/confirmations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
```
Comment (BandanaPandey) [inline app/controllers/users/confirmations_controller.rb]: @apeksha2697  please remove this file from PR

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+class UsersController < ApplicationController
+  before_action :authenticate_user!
+  def index
```
Comment (BandanaPandey) [inline app/controllers/users_controller.rb:4]: @apeksha2697  add one line spacing between before_action and index

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  end
+  
+  def destroy
+    @users = User.find(params[:id])
+    @users.destroy
+   
+    redirect_to users_path
+  end
+
+  private
```
Comment (BandanaPandey) [inline app/controllers/users_controller.rb:19]: @apeksha2697  please add one line spacing btw private and methods below

File: app/helpers/home_helper.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+module HomeHelper
```
Comment (BandanaPandey) [inline app/helpers/home_helper.rb:1]: @apeksha2697  no need to commit helpers file, if you are not adding any piece of code there

File: app/views/shared/_navbar.html.erb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    <div class="d-inline">
+      <a class="navbar-brand"><%=current_user.organisation.name + " | " + current_user.first_name + " " +current_user.last_name %></a>
+    </div>
+    <div class="form-inline">
+      <input class="form-control mr-sm-2" type="search" placeholder="Search" aria-label="Search">
+      <button class="btn btn-outline-success my-2 my-sm-0" type="submit">Search</button>
+      <div class="dropdown">
+        <button style="margin-left: 0.2cm;" class="btn btn-primary btn-circle btn-md" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><%=current_user.first_name.slice(0,1)+current_user.last_name.slice(0,1) %></button>
+        <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
+        <a class="dropdown-item" ><i class="fas fa-user-circle"></i><%=" " + current_user.first_name + " " +current_user.last_name %></a>
```
Comment (BandanaPandey) [inline app/views/shared/_navbar.html.erb]: @apeksha2697  please use string interpolation here, like this
"#{current_user.first_name}  #{current_user.last_name}"
or we can write a method in home helper, named full_name and inside that write this piece of code and can call that full_name method here

File: app/views/shared/_sidepanel.html.erb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+<% if user_signed_in? %>
+<div class="w3-sidebar w3-bar-block " style="width:6%">
+    <li><button class="btn btn-light btn-block"><%=link_to ('<i class="fas fa-2x fa-home"></i>').html_safe, root_path %></button></li>
```
Comment (BandanaPandey) [inline app/views/shared/_sidepanel.html.erb]: don't use html_safe, use sanitize function there
https://apidock.com/rails/ActionView/Helpers/SanitizeHelper/sanitize
Please read why we should not use html_safe and use sanitize

File: app/views/users/confirmations/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
+
+<%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "users/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email", value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email) %>
```
Comment (BandanaPandey) [inline app/views/users/confirmations/new.html.erb:8]: @apeksha2697  regarding value thing, use helper function, don't write logic in views.
Views should only be used for representation

File: app/views/users/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+      <th>Name</th>
+      <th>Email</th>
+      <th>Role</th>
+      <th></th>
+      <th></th>
+    </tr>
+    <% @users.all.each do |user| %>
+      <% if user.organisation_id == current_user.organisation_id %>
+        <tr>
+          <td><%= user.first_name + " " + user.last_name %></td>
```
Comment (BandanaPandey) [inline app/views/users/index.html.erb]: use helper_method please instead on concatenating first_name and last_name here

File: app/views/users/passwords/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,22 @@
+<%= stylesheet_link_tag 'style.css' %><br/>
+
+<body style="padding-left: 13cm;padding-right: 13cm;padding-top: 0.5cm">
+
+
```
Comment (BandanaPandey) [inline app/views/users/passwords/new.html.erb]: @apeksha2697  why there is so much space here ?

**Review by BandanaPandey — COMMENTED** [review]
@apeksha2697  Also please add testcases

File: app/controllers/home_controller.rb (hunk 1 above)
Comment (apeksha2697) [reply app/controllers/home_controller.rb:4]: done

File: app/controllers/users/sessions_controller.rb (hunk 3 above)
Comment (apeksha2697) [reply app/controllers/users/sessions_controller.rb]: done

File: app/controllers/users/registrations_controller.rb (hunk 4 above)
Comment (apeksha2697) [reply app/controllers/users/registrations_controller.rb:1]: done

File: app/controllers/users/passwords_controller.rb (hunk 5 above)
Comment (apeksha2697) [reply app/controllers/users/passwords_controller.rb]: done

File: app/controllers/users/confirmations_controller.rb (hunk 7 above)
Comment (apeksha2697) [reply app/controllers/users/confirmations_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 8 above)
Comment (apeksha2697) [reply app/controllers/users_controller.rb:4]: done

File: app/controllers/users_controller.rb (hunk 9 above)
Comment (apeksha2697) [reply app/controllers/users_controller.rb:19]: done

File: app/helpers/home_helper.rb (hunk 10 above)
Comment (apeksha2697) [reply app/helpers/home_helper.rb:1]: done

File: app/views/shared/_sidepanel.html.erb (hunk 12 above)
Comment (apeksha2697) [reply app/views/shared/_sidepanel.html.erb]: done

File: app/views/users/index.html.erb (hunk 14 above)
Comment (apeksha2697) [reply app/views/users/index.html.erb]: done

**Review by BandanaPandey — APPROVED** [review]

**Pull request merged**